std::lcm(UINT32_MAX, UINT32_MAX) fails when _LIBCPP_HARDENING_MODE != NONE

With any hardening mode above `_LIBCPP_HARDENING_MODE_NONE`, attempting to calculate `std::lcm` of various unsigned MAX values fails. For example with `UINT32_MAX`:

```c++
#include <limits>
#include <numeric>

static_assert(std::lcm(UINT32_MAX - 1, UINT32_MAX - 1) == UINT32_MAX - 1);

static_assert(std::lcm(UINT32_MAX, UINT32_MAX) == UINT32_MAX);

static_assert(std::lcm(INT32_MAX, INT32_MAX) == INT32_MAX);

#if 0 // expected to fail
static_assert(std::lcm(INT32_MIN, INT32_MIN) == INT32_MAX);
#endif
```

resulting in:

```text
lcmtest.cpp:6:15: error: static assertion expression is not an integral constant expression
    6 | static_assert(std::lcm(UINT32_MAX, UINT32_MAX) == UINT32_MAX);
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/v1/__numeric/gcd_lcm.h:80:3: note: subexpression not valid in a constant expression
   80 |   _LIBCPP_ASSERT_ARGUMENT_WITHIN_DOMAIN((numeric_limits<_Rp>::max() / __val1 > __val2), "Overflow in lcm");
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/v1/__config:379:74: note: expanded from macro '_LIBCPP_ASSERT_ARGUMENT_WITHIN_DOMAIN'
  379 | #    define _LIBCPP_ASSERT_ARGUMENT_WITHIN_DOMAIN(expression, message)   _LIBCPP_ASSERT(expression, message)
      |                                                                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/v1/__assert:23:10: note: expanded from macro '_LIBCPP_ASSERT'
   23 |        : _LIBCPP_ASSERTION_HANDLER(__FILE__ ":" _LIBCPP_TOSTRING(__LINE__) ": assertion " _LIBCPP_TOSTRING(            \
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   24 |              expression) " failed: " message "\n"))
      |              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/v1/__assertion_handler:27:62: note: expanded from macro '_LIBCPP_ASSERTION_HANDLER'
   27 | #  define _LIBCPP_ASSERTION_HANDLER(message) ((void)message, __builtin_trap())
      |                                                              ^~~~~~~~~~~~~~
lcmtest.cpp:6:15: note: in call to 'lcm<unsigned int, unsigned int>(4294967295, 4294967295)'
    6 | static_assert(std::lcm(UINT32_MAX, UINT32_MAX) == UINT32_MAX);
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lcmtest.cpp:8:15: error: static assertion expression is not an integral constant expression
    8 | static_assert(std::lcm(INT32_MAX, INT32_MAX) == INT32_MAX);
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/v1/__numeric/gcd_lcm.h:80:3: note: subexpression not valid in a constant expression
   80 |   _LIBCPP_ASSERT_ARGUMENT_WITHIN_DOMAIN((numeric_limits<_Rp>::max() / __val1 > __val2), "Overflow in lcm");
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/v1/__config:379:74: note: expanded from macro '_LIBCPP_ASSERT_ARGUMENT_WITHIN_DOMAIN'
  379 | #    define _LIBCPP_ASSERT_ARGUMENT_WITHIN_DOMAIN(expression, message)   _LIBCPP_ASSERT(expression, message)
      |                                                                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/v1/__assert:23:10: note: expanded from macro '_LIBCPP_ASSERT'
   23 |        : _LIBCPP_ASSERTION_HANDLER(__FILE__ ":" _LIBCPP_TOSTRING(__LINE__) ": assertion " _LIBCPP_TOSTRING(            \
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   24 |              expression) " failed: " message "\n"))
      |              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/v1/__assertion_handler:27:62: note: expanded from macro '_LIBCPP_ASSERTION_HANDLER'
   27 | #  define _LIBCPP_ASSERTION_HANDLER(message) ((void)message, __builtin_trap())
      |                                                              ^~~~~~~~~~~~~~
lcmtest.cpp:8:15: note: in call to 'lcm<int, int>(2147483647, 2147483647)'
    8 | static_assert(std::lcm(INT32_MAX, INT32_MAX) == INT32_MAX);
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2 errors generated.
```

And similar for `UINT64_MAX`, `std::numeric_limits<unsigned>::max()`, etc.

For signed quantities this could indeed occur, so for example `std::lcm(INT32_MIN, INT32_MIN)` is undefined. But MAX values, which are positive, should still work, even for signed types.

In libstdc++ this is handled at compile time by doing:

```c++
      if constexpr (is_signed_v<_Ct>)
        if (__is_constant_evaluated())
          return __r * __n2; // constant evaluation can detect overflow here.

      bool __overflow = __builtin_mul_overflow(__r, __n2, &__r);
```

but libc++ does:

```c++
  using _Rp  = common_type_t<_Tp, _Up>;
  _Rp __val1 = __ct_abs<_Rp, _Tp>()(__m) / std::gcd(__m, __n);
  _Rp __val2 = __ct_abs<_Rp, _Up>()(__n);
  _LIBCPP_ASSERT_ARGUMENT_WITHIN_DOMAIN((numeric_limits<_Rp>::max() / __val1 > __val2), "Overflow in lcm");
  return __val1 * __val2;
```

I think the assertion should check for `>=`, in case the types (or the common type at least) are unsigned. For the other cases, `>` should be fine. (This avoids using `__builtin_mul_overflow` which may not be available everywhere, but I'm not sure if it isn't used anywhere else.)

--- a/libcxx/include/__numeric/gcd_lcm.h
+++ b/libcxx/include/__numeric/gcd_lcm.h
@@ -123,7 +123,7 @@ _LIBCPP_CONSTEXPR _LIBCPP_HIDE_FROM_ABI common_type_t<_Tp, _Up> lcm(_Tp __m, _Up
   return __res;
 }
 
-#endif // _LIBCPP_STD_VER
+#endif // _LIBCPP_STD_VER >= 17
 
 _LIBCPP_END_NAMESPACE_STD
 

--- a/libcxx/include/__numeric/gcd_lcm.h
+++ b/libcxx/include/__numeric/gcd_lcm.h
@@ -118,8 +118,7 @@ _LIBCPP_CONSTEXPR _LIBCPP_HIDE_FROM_ABI common_type_t<_Tp, _Up> lcm(_Tp __m, _Up
   _Rp __val1 = __ct_abs<_Rp, _Tp>()(__m) / std::gcd(__m, __n);
   _Rp __val2 = __ct_abs<_Rp, _Up>()(__n);
   _Rp __res;
-  bool __overflow = __builtin_mul_overflow(__val1, __val2, &__res);
-  (void)__overflow;
+  [[maybe_unused]] bool __overflow = __builtin_mul_overflow(__val1, __val2, &__res);
   _LIBCPP_ASSERT_ARGUMENT_WITHIN_DOMAIN(!__overflow, "Overflow in lcm");
   return __res;
 }

--- a/libcxx/include/__numeric/gcd_lcm.h
+++ b/libcxx/include/__numeric/gcd_lcm.h
@@ -117,8 +117,11 @@ _LIBCPP_CONSTEXPR _LIBCPP_HIDE_FROM_ABI common_type_t<_Tp, _Up> lcm(_Tp __m, _Up
   using _Rp  = common_type_t<_Tp, _Up>;
   _Rp __val1 = __ct_abs<_Rp, _Tp>()(__m) / std::gcd(__m, __n);
   _Rp __val2 = __ct_abs<_Rp, _Up>()(__n);
-  _LIBCPP_ASSERT_ARGUMENT_WITHIN_DOMAIN((numeric_limits<_Rp>::max() / __val1 > __val2), "Overflow in lcm");
-  return __val1 * __val2;
+  _Rp __res;
+  bool __overflow = __builtin_mul_overflow(__val1, __val2, &__res);
+  (void)__overflow;
+  _LIBCPP_ASSERT_ARGUMENT_WITHIN_DOMAIN(!__overflow, "Overflow in lcm");
+  return __res;
 }
 
 #endif // _LIBCPP_STD_VER

--- a/libcxx/test/std/numerics/numeric.ops/numeric.ops.lcm/lcm.pass.cpp
+++ b/libcxx/test/std/numerics/numeric.ops/numeric.ops.lcm/lcm.pass.cpp
@@ -89,6 +89,13 @@ constexpr bool do_test(int = 0)
     return accumulate;
 }
 
+template <class T>
+constexpr bool test_limits() {
+    assert(std::lcm(std::numeric_limits<T>::max() - 1, std::numeric_limits<T>::max() - 1) == std::numeric_limits<T>::max() - 1);
+    assert(std::lcm(std::numeric_limits<T>::max(), std::numeric_limits<T>::max()) == std::numeric_limits<T>::max());
+    return true;
+}
+
 int main(int argc, char**)
 {
     int non_cce = argc; // a value that can't possibly be constexpr
@@ -141,5 +148,22 @@ int main(int argc, char**)
     assert(res1 == 1324997410816LL);
     }
 
-  return 0;
+    // https://github.com/llvm/llvm-project/issues/96196
+    {
+        assert(test_limits<unsigned int>());
+        assert(test_limits<std::uint32_t>());
+        assert(test_limits<std::uint64_t>());
+        assert(test_limits<int>());
+        assert(test_limits<std::int32_t>());
+        assert(test_limits<std::int64_t>());
+
+        static_assert(test_limits<unsigned int>(), "");
+        static_assert(test_limits<std::uint32_t>(), "");
+        static_assert(test_limits<std::uint64_t>(), "");
+        static_assert(test_limits<int>(), "");
+        static_assert(test_limits<std::int32_t>(), "");
+        static_assert(test_limits<std::int64_t>(), "");
+    }
+
+    return 0;
 }

--- a/libcxx/test/std/numerics/numeric.ops/numeric.ops.lcm/lcm.pass.cpp
+++ b/libcxx/test/std/numerics/numeric.ops/numeric.ops.lcm/lcm.pass.cpp
@@ -16,6 +16,7 @@
 #include <cassert>
 #include <climits>
 #include <cstdint>
+#include <limits>
 #include <type_traits>
 #include "test_macros.h"
 

